### PR TITLE
Treat conditional code as if it was always active

### LINKED ISSUE #5574: unneeded_synthesized_initializer removes needed initializers if there is any other initializer enclosed within conditional macro
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

```swift
struct Foo {
    var bar: String

   // warning or removed according to unneeded_synthesized_initializer
    init(bar: String) {
        self.bar = bar
    }

    // or any other macros, such as canImport
    #if DEBUG  
    init(data: CGFloat) {
        self.bar = "\(data)"
    }
    #endif
}

/// compile error if autocorrect
let foo = Foo(bar: "bar")
```

### Environment

* SwiftLint version: 0.55.0   /  0.53.0
* Installation method used:  CocoaPods  / Homebrew

### PR DESCRIPTION
**Pull request by SimplyDanny**
Fixes #5574.

### CODE REVIEW

**Comment by SimplyDanny** [comment]
I'd appreciate your feedback on this @mildm8nnered as you have provided the initial implementation of this rule. This is the last issue reported after the 0.55.0 release. Once it's merged, I may cut 0.55.1.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift [hunk 1]
```diff
@@ -71,42 +71,51 @@ private extension UnneededSynthesizedInitializerRule {
     }
 }
 
+private final class ElementCollector: SyntaxAnyVisitor {
+    var initializers = [InitializerDeclSyntax]()
+    var varDecls = [VariableDeclSyntax]()
+
+    override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+        if node.isProtocol((any NamedDeclSyntax).self) {
```
Comment (mildm8nnered) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift:79]: ternary operator and save on the `return`s here potentially

**Review by mildm8nnered — APPROVED** [review]
lgtm

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift [hunk 2]
```diff
@@ -71,42 +71,51 @@ private extension UnneededSynthesizedInitializerRule {
...
+            return .skipChildren
+        }
+        return .visitChildren
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnneededSynthesizedInitializerRule.swift:82]:
```suggestion
        node.isProtocol((any NamedDeclSyntax).self) ? .skipChildren : .visitChildren
```

**Pull request merged**